File: Leetcode/Heap (Priority Queue)/826. Most Profit Assigning Work.cpp
class Solution {
private:
    struct Job {
        int d, p;
        Job(int difficulty, int profit): d(difficulty), p(profit) {}
    };

public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        auto compare = [](const Job& j1, const Job& j2) {
            return j1.d > j2.d;
        };

        priority_queue<Job, vector<Job>, decltype(compare)> jobs;
        for(int i = 0; i < profit.size(); i++) jobs.push(Job(difficulty[i], profit[i]));

        sort(worker.begin(), worker.end());

        int mxProfit = 0, mxCurrent = 0;
        for(auto w: worker) {
            while(!jobs.empty() and w >= jobs.top().d) {
                mxCurrent = max(mxCurrent, jobs.top().p);
                jobs.pop();
            }
            mxProfit += mxCurrent;
        }

        return mxProfit;
    }
};

File: Leetcode/Heap (Priority Queue)/826. Most Profit Assigning Work.py
class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        jobs = sorted(zip(difficulty, profit))
        worker.sort()
        
        mxProfit = mxCurrent = pos = 0
        for w in worker:
          while pos < len(jobs) and w >= jobs[pos][0]:
            mxCurrent = max(mxCurrent, jobs[pos][1])
            pos += 1
          mxProfit += mxCurrent
          
        return mxProfit